### ISSUE TITLE
Const numbers that index into tuples do not produce the correct type

### ISSUE THREAD

**Issue by sebdoucet** [microsoft/TypeScript#6678]
Hi,

Constants aren't evaluate to access and cast properly union types values:

``` Typescript
const ID = 0;
const NAME = 1;

let data: [number, string] = [0, "bob"];

// This need cast to compile
let id: number = data[ID];
let name: string = data[NAME];
```

Is there a reason why it's not supported ?

Regards,
Sébastien

**Comment by RyanCavanaugh** [comment]
We did this for const enums (#3411); I don't see why we wouldn't for this case either

**Comment by sebdoucet** [comment]
It's not compile because you need a cast, it does not recognize ID as 0, so it does not detect you try to get the first entry of the union types, so it does not cast it.

What is important to read is the tuple have first entry as number and second entry as string. Without const, data[0] return a number and data[1]  return a string. With const it does not work.

**Comment by RyanCavanaugh** [comment]
Accepting PRs.

See also linked issues #6804 and #6805

**Comment by DrSammyD** [comment]
```
class x{
   public y(){ return "string" }
}
var z= new x;
const y = "y";
var w = z["y"]();
var v = z[y]();
v.length; //No longer type safe
w.length; //Type safe
```
This too?

**Comment by mhegazy** [comment]
> Should this be changed too?

i do not see why not.

**Comment by aleksey-bykov** [comment]
got bitten by it, please fix asap:

nightly build Oct 24, 2016

```typescript
const values: [number, string] = [1, 'a'];

const one = values[0]; // number, OK

const index = 0; // index is `0` (not just a number)
const another = values[index]; // expected number, actual: string | number <-- WHY?
```

**Comment by JoshuaKGoldberg** [comment]
@RyanCavanaugh this seems to be fixed when checking TypeScript 3.3.3. `id` and `name` from the OP are correctly `number` and `string` with and without their explicit type annotations.
